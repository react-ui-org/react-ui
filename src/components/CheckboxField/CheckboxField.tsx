import React, {
  useContext,
  forwardRef,
} from 'react';
import { withGlobalProps } from '../../providers/globalProps';
import { classNames } from '../../helpers/classNames/classNames';
import { transferProps } from '../../helpers/transferProps';
import { getRootValidationStateClassName } from '../_helpers/getRootValidationStateClassName';
import { FormLayoutContext } from '../FormLayout';
import type { CheckboxFieldProps } from './CheckboxField.types';
import styles from './CheckboxField.module.scss';

export const CheckboxField = forwardRef<HTMLInputElement, CheckboxFieldProps>((props, ref) => {
  const {
    disabled = false,
    helpText,
    id,
    isLabelVisible = true,
    label,
    labelPosition = 'after',
    renderAsRequired = false,
    required = false,
    validationState,
    validationText,
    ...restProps
  } = props;
  const context = useContext(FormLayoutContext);

  return (
    <label
      className={classNames(
        styles.root,
        context && styles.isRootInFormLayout,
        context && context.layout === 'horizontal' ? styles.isRootLayoutHorizontal : styles.isRootLayoutVertical,
        labelPosition === 'before' && styles.hasRootLabelBefore,
        disabled && styles.isRootDisabled,
        (renderAsRequired || required) && styles.isRootRequired,
        getRootValidationStateClassName(validationState, styles),
      )}
      htmlFor={id}
      id={id && `${id}__label`}
    >
      <div className={styles.field}>
        <div
          className={classNames(
            styles.label,
            !isLabelVisible && styles.isLabelHidden,
          )}
          id={id && `${id}__labelText`}
        >
          {label}
        </div>
        <input
          {...transferProps(restProps)}
          className={styles.input}
          disabled={disabled}
          id={id}
          ref={ref}
          required={required}
          type="checkbox"
        />
      </div>
      {helpText && (
        <div
          className={styles.helpText}
          id={id && `${id}__helpText`}
        >
          {helpText}
        </div>
      )}
      {validationText && (
        <div
          className={styles.validationText}
          id={id && `${id}__validationText`}
        >
          {validationText}
        </div>
      )}
    </label>
  );
});

export const CheckboxFieldWithGlobalProps = withGlobalProps(CheckboxField, 'CheckboxField');

export default CheckboxFieldWithGlobalProps;
