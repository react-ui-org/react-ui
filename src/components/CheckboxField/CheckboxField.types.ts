import { ReactNode } from 'react';
import {
  Position,
  Validation,
} from '../../types';

export type CheckboxFieldProps = React.ComponentProps<'input'> & {
  /**
   * If `true`, the input will be disabled.
   */
  disabled?: boolean;
  /**
   * Optional help text.
   */
  helpText?: ReactNode;
  /**
   * ID of the `<input>` HTML element.
   *
   * Also serves as base for ids of nested elements:
   * * `<ID>__label`
   * * `<ID>__labelText`
   * * `<ID>__helpText`
   * * `<ID>__validationText`
   */
  id?: string;
  /**
   * If `false`, the label will be visually hidden (but remains accessible by assistive
   * technologies).
   */
  isLabelVisible?: boolean;
  /**
   * Checkbox field label.
   */
  label: ReactNode;
  /**
   * Placement of the label relative to the input.
   */
  labelPosition?: Position;
  /**
   * If `true`, the input will be rendered as if it was required.
   */
  renderAsRequired?: boolean;
  /**
   * If `true`, the input will be made and rendered as required, regardless of the `renderAsRequired` prop.
   */
  required?: boolean;
  /**
   * Alter the field to provide feedback based on validation result.
   */
  validationState?: Validation;
  /**
   * Validation message to be displayed.
   */
  validationText?: ReactNode;
};
