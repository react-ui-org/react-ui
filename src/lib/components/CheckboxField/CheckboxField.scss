@use "../../styles/tools/form-fields/foundation";
@use "../../styles/tools/form-fields/inline-field-elements";
@use "../../styles/tools/form-fields/inline-field-layout";
@use "../../styles/tools/form-fields/variants";
@use "../../styles/tools/accessibility";

// Foundation
.root {
    @include foundation.root();
    @include inline-field-layout.root();
    @include inline-field-elements.min-tap-target($type: checkbox);
    @include variants.visual(check);
}

.label {
    @include foundation.label();
}

.field {
    @include inline-field-layout.field();
}

.input {
    @include inline-field-elements.check-input($type: checkbox);
}

.helpText,
.validationText {
    @include foundation.help-text();
}

.isRootRequired .label {
    @include foundation.label-required();
}

// States
.isRootStateInvalid {
    @include variants.validation(invalid);
}

.isRootStateValid {
    @include variants.validation(valid);
}

.isRootStateWarning {
    @include variants.validation(warning);
}

// Invisible label
.isLabelHidden {
    @include accessibility.hide-text();
}

// Layouts
.hasRootLabelBefore {
    @include inline-field-layout.has-label-before();
}

.isRootInFormLayout {
    @include inline-field-layout.in-form-layout();
}
