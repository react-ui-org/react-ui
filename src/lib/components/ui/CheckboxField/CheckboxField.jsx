import PropTypes from 'prop-types';
import React from 'react';
import getRootValidationStateClassName from '../../../helpers/getRootValidationStateClassName';
import transferProps from '../../../utils/transferProps';
import withForwardedRef from '../withForwardedRef';
import styles from './CheckboxField.scss';

export const CheckboxField = (props) => {
  const {
    changeHandler,
    checked,
    disabled,
    forwardedRef,
    helpText,
    id,
    inFormLayout,
    isLabelVisible,
    label,
    labelPosition,
    layout,
    required,
    validationState,
    validationText,
    value,
  } = props;

  const propsToTransfer = transferProps(
    props,
    ['changeHandler', 'checked', 'disabled', 'helpText', 'id', 'inFormLayout', 'isLabelVisible',
      'label', 'labelPosition', 'layout', 'required', 'validationState', 'validationText', 'value'],
  );

  return (
    <div
      className={[
        styles.root,
        inFormLayout ? styles.isRootInFormLayout : '',
        layout === 'horizontal' ? styles.isRootLayoutHorizontal : '',
        getRootValidationStateClassName(validationState, styles),
      ].join(' ')}
    >
      <label
        className={[
          styles.inputWrap,
          labelPosition === 'before' ? styles.labelPositionBefore : styles.labelPositionAfter,
        ].join(' ')}
        htmlFor={id}
        id={`${id}__label`}
      >
        <input
          {...propsToTransfer}
          id={id}
          value={value}
          onChange={changeHandler}
          disabled={disabled}
          checked={checked}
          ref={forwardedRef}
          required={required}
          type="checkbox"
          className={styles.input}
        />
        <span className={styles.label}>
          <span
            className={[
              styles.labelInner,
              isLabelVisible ? '' : styles.isLabelHidden,
            ].join(' ')}
            id={`${id}__labelText`}
          >
            {label}
          </span>
        </span>
      </label>
      {helpText && (
        <div
          className={styles.helpText}
          id={`${id}__helpText`}
        >
          {helpText}
        </div>
      )}
      {validationText && (
        <div
          className={styles.validationText}
          id={`${id}__validationText`}
        >
          {validationText}
        </div>
      )}
    </div>
  );
};

CheckboxField.defaultProps = {
  changeHandler: null,
  checked: false,
  disabled: false,
  forwardedRef: undefined,
  helpText: null,
  inFormLayout: false,
  isLabelVisible: true,
  labelPosition: 'after',
  layout: 'vertical',
  required: false,
  validationState: null,
  validationText: null,
  value: undefined,
};

CheckboxField.propTypes = {
  /**
   * Function to call when the input is toggled.
   */
  changeHandler: PropTypes.func,
  /**
   * If `true`, the input will be checked.
   */
  checked: PropTypes.bool,
  /**
   * If `true`, the input will be disabled.
   */
  disabled: PropTypes.bool,
  /**
   * Reference forwarded to the `input` element.
   */
  forwardedRef: PropTypes.oneOfType([
    PropTypes.func,
    PropTypes.shape({ current: PropTypes.any }),
  ]),
  /**
   * Optional help text.
   */
  helpText: PropTypes.node,
  /**
   * ID of the input HTML element. It also serves as a prefix for important inner elements:
   * `<ID>__label`, `<ID>__labelText`, `<ID>__helpText`, and `<ID>__validationText`.
   */
  id: PropTypes.string.isRequired,
  /**
   * Treat the field differently when it's inside a FormLayout. Do not set manually!
   */
  inFormLayout: PropTypes.bool,
  /**
   * If `false`, the label will be visually hidden (but remains accessible by assistive
   * technologies).
   */
  isLabelVisible: PropTypes.bool,
  /**
   * Checkbox field label.
   */
  label: PropTypes.string.isRequired,
  /**
   * Placement of the label relative to the input.
   */
  labelPosition: PropTypes.oneOf(['before', 'after']),
  /**
   * Layout of the field. It has impact only on CheckboxField inside a FormLayout.
   */
  layout: PropTypes.oneOf(['horizontal', 'vertical']),
  /**
   * If `true`, the input will be required.
   */
  required: PropTypes.bool,
  /**
   * Alter the field to provide feedback based on validation result.
   */
  validationState: PropTypes.oneOf(['invalid', 'valid', 'warning']),
  /**
   * Validation message to be displayed.
   */
  validationText: PropTypes.node,
  /**
   * Value of the input.
   */
  value: PropTypes.oneOfType([
    PropTypes.string,
    PropTypes.number,
  ]),
};

export default withForwardedRef(CheckboxField);
