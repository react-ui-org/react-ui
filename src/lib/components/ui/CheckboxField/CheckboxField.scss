@import '../../../styles/settings/borders';
@import '../../../styles/settings/forms.deprecated';
@import '../../../styles/tools/accessibility';
@import '../../../styles/tools/forms/foundation';
@import '../../../styles/tools/forms.deprecated';

.root {
  position: relative;
  display: flex;
  flex-direction: column;
  box-sizing: border-box;
}

.inputWrap {
  position: relative;
  min-height: $checkbox-radio-label-height;
}

.input {
  @include form-checkbox-radio-input-styles($checkbox-radio-size, 0);
}

.label {
  @include form-field-label();
  @include form-checkbox-radio-label-styles($checkbox-radio-label-height, $checkbox-radio-padding-left);
}

.input + .label::before {
  @include form-checkbox-radio-shared-inactive(
    $checkbox-radio-size,
    $checkbox-radio-top,
    0,
    $form-field-border-color
  );

  border-radius: $border-radius;
}

.input + .label::after {
  content: '';
  position: absolute;
  top: 50%;
  left: calc(#{$checkbox-checkmark-size / 2} + 1px);
  display: none;
  width: $checkbox-checkmark-size;
  height: calc((#{$checkbox-checkmark-size} / 2) + 1px);
  border-color: $checkbox-checkmark-color;
  border-bottom: $checkbox-checkmark-border-style;
  border-left: $checkbox-checkmark-border-style;
  transform: translateY(-50%) rotate(-45deg);
}

// stylelint-disable selector-max-specificity, selector-nested-pattern
.labelPositionBefore {
  .input + .label::before {
    right: 0;
    left: auto;
  }

  .input + .label::after {
    right: calc(#{$checkbox-checkmark-size / 2} + 1px);
    left: auto;
  }

  .label {
    padding-right: $checkbox-radio-padding-left;
    padding-left: 0;
  }
}

.labelPositionAfter {
  .input + .label::before {
    left: 0;
  }

  .input + .label::after {
    right: 0;
  }
}

.input:disabled + .label {
  @include form-disabled-state();

  color: $checkbox-disabled-text-color;
}

.input:checked + .label::before {
  background: $form-field-background-color;
}

.input:checked + .label::after {
  display: inline-block;
}

.input:indeterminate + .label::before {
  border: $border-width solid $checkbox-border-color-indeterminate;
}

.input:indeterminate + .label::after {
  display: inline-block;
  border-bottom-color: $checkbox-border-color-indeterminate;
  border-left: 0;
  transform: translate(0, ($checkbox-checkmark-size / 2));
}

.input:focus + .label::before {
  outline: 0;
  box-shadow: $radio-focused-box-shadow;
}

.input:disabled + .label::before {
  background-color: $checkbox-disabled-background-color;
}

.input:focus:checked + .label::before {
  outline: 0;
  box-shadow: $radio-focused-box-shadow;
}

.input:disabled:checked + .label::before {
  background-color: $radio-disabled-background-color;
  box-shadow: $radio-disabled-checked-box-shadow;
}

.input:required + .label .labelInner::after {
  content: ' *';
}

.input:checked:disabled + .label::after {
  $disabled-border-style: ($checkbox-radio-size / 8) solid $checkbox-disabled-checkmark-color;

  border-bottom: $disabled-border-style;
  border-left: $disabled-border-style;
}

.helpText,
.validationText {
  @include form-field-help-text();
}

.validationText {
  min-height: 1rem;
}

.isLabelHidden {
  @include accessibility-hide-text();
}

// States
// stylelint-disable no-descending-specificity
.isRootStateInvalid .label,
.isRootStateInvalid .validationText {
  color: $form-invalid-color;
}

.isRootStateInvalid .input + .label::before,
.isRootStateInvalid .input:checked + .label::before {
  border-color: $form-invalid-color;
}

.isRootStateInvalid .input:checked + .label::before {
  background-color: $form-invalid-color;
}

.isRootStateValid .label,
.isRootStateValid .validationText {
  color: $form-valid-color;
}

.isRootStateValid .input + .label::before,
.isRootStateValid .input:checked + .label::before {
  border-color: $form-valid-color;
}

.isRootStateValid .input:checked + .label::before {
  background-color: $form-valid-color;
}

.isRootStateWarning .label,
.isRootStateWarning .validationText {
  color: $form-warning-color;
}

.isRootStateWarning .input + .label::before,
.isRootStateWarning .input:checked + .label::before {
  border-color: $form-warning-color;
}

.isRootStateWarning .input:checked + .label::before {
  background-color: $form-warning-color;
}

.isRootLayoutHorizontal {
  position: relative; // Just set something until the component is refactored (#20).
}

.isRootInFormLayout {
  @include form-field-in-form-layout();
}
