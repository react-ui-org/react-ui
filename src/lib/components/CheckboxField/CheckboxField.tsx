import React, { useContext } from 'react';
import { withGlobalProps } from '../../provider';
import { classNames } from '../../utils/classNames';
import { getRootValidationStateClassName } from '../_helpers/getRootValidationStateClassName';
import { transferProps } from '../_helpers/transferProps';
import { FormLayoutContext } from '../FormLayout';
import styles from './CheckboxField.scss';
import { CheckboxFieldProps } from './CheckboxField.types';

export const CheckboxField: React.FunctionComponent<CheckboxFieldProps> = React.forwardRef<HTMLInputElement,
CheckboxFieldProps>(({
  disabled = false,
  helpText,
  id,
  isLabelVisible = true,
  label,
  labelPosition = 'after',
  required = false,
  validationState,
  validationText,
  ...restProps
}, ref) => {
  const context = useContext(FormLayoutContext);

  return (
    <label
      className={classNames(
        styles.root,
        context && styles.isRootInFormLayout,
        context && context.layout === 'horizontal' ? styles.isRootLayoutHorizontal : styles.isRootLayoutVertical,
        labelPosition === 'before' && styles.hasRootLabelBefore,
        disabled && styles.isRootDisabled,
        required && styles.isRootRequired,
        getRootValidationStateClassName(validationState, styles),
      )}
      htmlFor={id}
      id={id && `${id}__label`}
    >
      <div className={styles.field}>
        <div
          className={classNames(
            styles.label,
            !isLabelVisible && styles.isLabelHidden,
          )}
          id={id && `${id}__labelText`}
        >
          {label}
        </div>
        <input
          {...transferProps(restProps)}
          className={styles.input}
          disabled={disabled}
          id={id}
          ref={ref}
          required={required}
          type="checkbox"
        />
      </div>
      {helpText && (
        <div
          className={styles.helpText}
          id={id && `${id}__helpText`}
        >
          {helpText}
        </div>
      )}
      {validationText && (
        <div
          className={styles.validationText}
          id={id && `${id}__validationText`}
        >
          {validationText}
        </div>
      )}
    </label>
  );
});

export const CheckboxFieldWithGlobalProps = withGlobalProps(CheckboxField, 'CheckboxField');

export default CheckboxFieldWithGlobalProps;
