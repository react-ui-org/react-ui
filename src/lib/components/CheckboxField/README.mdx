---
name: CheckboxField
menu: 'Inputs'
route: /components/checkbox-field
---

# CheckboxField

CheckboxField allows users to select one or more options from a set.

import {
  Playground,
  Props,
} from 'docz'
import { CheckboxField } from './CheckboxField'

## Basic Usage

To implement the CheckboxField component, you need to import it first:

```js
import { CheckboxField } from '@react-ui-org/react-ui';
```

And use it:

<Playground>
  {() => {
    const [agree, setAgree] = React.useState(true);
    return (
      <CheckboxField
        checked={agree}
        label="I agree"
        onChange={() => setAgree(!agree)}
      />
    );
  }}
</Playground>

See [API](#api) for all available options.

## General Guidelines

- Use the CheckboxField when there are lists of options and the user may
  **select any number of choices,** including zero, one, or several. For
  selecting just a single option from the list, use either the
  [Radio](/components/radio) or [SelectField](/components/select-field)
  component.

- Use the CheckboxField for boolean (true/false) input **in forms that require
  to be confirmed** by a button to become active. To toggle things on or off
  with an immediate effect (without confirmation), use rather the
  [Toggle](/components/toggle) component.

- **Use positive wording for the checkbox label,** so that it's clear what will
  happen when the user turns on the checkbox. Avoid negations such as ‚ÄúDon't
  send me any emails‚Äù which would mean that the user needs to turn the
  checkbox **on** in order for something **not** to happen.

- Only make the CheckboxField's label invisible when there is **another visual
  clue** to guide users through toggling the input.

- When a short label is not enough, use **help texts to guide users** before
  they enter anything.

- Use **clear, calm error messages** when there's a problem with what they
  entered.

- **Visually present groups of choices as groups,** and clearly separate them
  from other groups on the same page.

üìñ [Read more about checkboxes and radios at Nielsen Norman Group.](https://www.nngroup.com/articles/checkboxes-vs-radio-buttons/)

## Help Text

You may provide an additional help text to clarify what will be the impact of
turning the checkbox on or off.

<Playground>
  {() => {
    const [getNewsletter, setGetNewsletter] = React.useState(true);
    return (
      <CheckboxField
        checked={getNewsletter}
        helpText="We will not bother you more than once a month, we promise!"
        label="Send me newsletter"
        onChange={() => setGetNewsletter(!getNewsletter)}
      />
    );
  }}
</Playground>

## Label Options

In some cases, it may be convenient to visually hide the checkbox label. The
label remains accessible to assistive technologies.

<Playground>
  {() => {
    const [checked, setChecked] = React.useState(true);
    return (
      <CheckboxField
        checked={checked}
        isLabelVisible={false}
        label="You cannot see this"
        onChange={() => setChecked(!checked)}
      />
    );
  }}
</Playground>

It's also possible to display label before input:

<Playground>
  {() => {
    const [checked, setChecked] = React.useState(true);
    return (
      <CheckboxField
        checked={checked}
        label="Label precedes input"
        labelPosition="before"
        onChange={() => setChecked(!checked)}
      />
    );
  }}
</Playground>

## States

### Validation States

Validation states visually present the result of validation of the input. You
should always **provide validation message for states other than valid** so
users know what happened and what action they should take or what options they
have.

<Playground>
  {() => {
    const [agree, setAgree] = React.useState(true);
    return (
      <>
        <CheckboxField
          checked={agree}
          label="I have read and agree with terms and conditions"
          onChange={() => setAgree(!agree)}
          validationState="valid"
        />
        <CheckboxField
          checked={agree}
          label="I have read and agree with terms and conditions"
          onChange={() => setAgree(!agree)}
          validationState="warning"
          validationText="Please wait 10 minutes until we verify your data."
        />
        <CheckboxField
          checked={agree}
          label="I have read and agree with terms and conditions"
          onChange={() => setAgree(!agree)}
          required
          validationState="invalid"
          validationText="You must agree to be able to proceed."
        />
      </>
    );
  }}
</Playground>

### Disabled State

Disabled state makes the input unavailable.

<Playground>
  <CheckboxField
    disabled
    label="Disabled checkbox"
  />
  <CheckboxField
    checked
    disabled
    label="Disabled checkbox, checked"
  />
</Playground>

## API

In addition to the options below, you can specify [React synthetic events] or
any custom HTML attributes that do not interfere with the API, and they will be
passed to the `input` HTML element. This enables making the component
interactive and helps improve its accessibility.

<Props table of={CheckboxField} />

## Theming

Head to [Forms Theming](/customize/theming/forms) to see shared form theming
options. On top of that, the following options are available for CheckboxField.

| Custom Property                                                      | Description                                  |
|----------------------------------------------------------------------|----------------------------------------------|
| `--rui-FormField--check__input--checkbox__border-radius`             | Input corner radius                          |
| `--rui-FormField--check__input--checkbox--checked__background-image` | Background image of checked input            |

[React synthetic events]: https://reactjs.org/docs/events.html
