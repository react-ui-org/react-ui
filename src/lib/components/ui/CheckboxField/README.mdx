---
name: CheckboxField
menu: 'UI'
route: /components/ui/checkbox-field
---

# CheckboxField

CheckboxField allows users to select one or more options from a set.

import {
  Playground,
  Props,
} from 'docz'
import { CheckboxField } from './CheckboxField'

## Basic Usage

To implement the CheckboxField component, you need to import it first:

```js
import { CheckboxField } from '@react-ui-org/react-ui';
```

And use it:

<Playground>
  {() => {
    const [agree, setAgree] = React.useState(true);
    return (
      <CheckboxField
        changeHandler={() => setAgree(!agree)}
        checked={agree}
        id="my-checkbox"
        label="I agree"
      />
    );
  }}
</Playground>

See [API](#api) for all available options.

## General Guidelines

- Use the CheckboxField when there are lists of options and the user may
  **select any number of choices,** including zero, one, or several. For
  selecting just a single option from the list, use either the
  [Radio](/components/ui/radio) or [SelectField](/components/ui/select-field)
  component.

- Use the CheckboxField for boolean (true/false) input **in forms that require
  to be confirmed** by a button to become active. To toggle things on or off
  with an immediate effect (without confirmation), use rather the
  [Toggle](/components/ui/toggle) component.

- **Use positive wording for the checkbox label,** so that it's clear what will
  happen when the user turns on the checkbox. Avoid negations such as ‚ÄúDon't
  send me any emails‚Äù which would mean that the user needs to turn the
  checkbox **on** in order for something **not** to happen.

- Only make the CheckboxField's label invisible when there is **another visual
  clue** to guide users through toggling the input.

- When a short label is not enough, use **help texts to guide users** before
  they enter anything.

- Use **clear, calm error messages** when there's a problem with what they
  entered.

- **Visually present groups of choices as groups,** and clearly separate them
  from other groups on the same page.

üìñ [Read more about checkboxes and radios at Nielsen Norman Group.](https://www.nngroup.com/articles/checkboxes-vs-radio-buttons/)

## Help Text

You may provide an additional help text to clarify what will be the impact of
turning the checkbox on or off.

<Playground>
  {() => {
    const [getNewsletter, setGetNewsletter] = React.useState(true);
    return (
      <CheckboxField
        changeHandler={() => setGetNewsletter(!getNewsletter)}
        checked={getNewsletter}
        helpText="We will not bother you more than once a month, we promise!"
        id="my-checkbox-with-help-text"
        label="Send me newsletter"
      />
    );
  }}
</Playground>

## Label Options

In some cases, it may be convenient to visually hide the checkbox label. The
label remains accessible to assistive technologies.

<Playground>
  {() => {
    const [checked, setChecked] = React.useState(true);
    return (
      <CheckboxField
        changeHandler={() => setChecked(!checked)}
        checked={checked}
        id="my-checkbox-with-invisible-label"
        isLabelVisible={false}
        label="You cannot see this"
      />
    );
  }}
</Playground>

It's also possible to display label before input:

<Playground>
  {() => {
    const [checked, setChecked] = React.useState(true);
    return (
      <CheckboxField
        changeHandler={() => setChecked(!checked)}
        checked={checked}
        id="my-checkbox-with-label-before-input"
        label="Label precedes input"
        labelPosition="before"
      />
    );
  }}
</Playground>

## States

### Validation States

Validation states visually present the result of validation of the input. You
should always **provide validation message for states other than valid** so
users know what happened and what action they should take or what options they
have.

<Playground>
  {() => {
    const [agree, setAgree] = React.useState(true);
    return (
      <>
        <CheckboxField
          changeHandler={() => setAgree(!agree)}
          checked={agree}
          id="my-checkbox-valid"
          label="I have read and agree with terms and conditions"
          validationState="valid"
        />
        <CheckboxField
          changeHandler={() => setAgree(!agree)}
          checked={agree}
          id="my-checkbox-warning"
          label="I have read and agree with terms and conditions"
          validationState="warning"
          validationText="Please wait 10 minutes until we verify your data."
        />
        <CheckboxField
          changeHandler={() => setAgree(!agree)}
          checked={agree}
          id="my-checkbox-invalid"
          label="I have read and agree with terms and conditions"
          required
          validationState="invalid"
          validationText="You must agree to be able to proceed."
        />
      </>
    );
  }}
</Playground>

### Disabled State

Disabled state makes the input unavailable.

<Playground>
  <CheckboxField
    disabled
    id="my-checkbox-disabled"
    label="Disabled checkbox"
  />
  <CheckboxField
    checked
    disabled
    id="my-checkbox-checked-disabled"
    label="Disabled checkbox, checked"
  />
</Playground>

## API

In addition to the options below, you can add any custom attributes that do not
interfere with the API, and they will be passed to the `input` HTML element.
This is useful mainly to improve component's accessibility.

<Props table of={CheckboxField} />

## Theming

Head to [Forms Theming](/customize/theming/forms) to see shared form theming
options. On top of that, the following options are available for CheckboxField.

| Custom Property                                                    | Description                                    |
|--------------------------------------------------------------------|------------------------------------------------|
| `--rui-form-field-check-input-checkbox-border-radius`              | Input border radius                            |
| `--rui-form-field-check-input-checkbox-checked-background-image`  | Checked input background image (inline, URL, ‚Ä¶) |
