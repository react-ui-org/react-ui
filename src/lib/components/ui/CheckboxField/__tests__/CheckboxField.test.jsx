import React from 'react';
import {
  shallow,
  mount,
} from 'enzyme';
import { shallowToJson } from 'enzyme-to-json';
import sinon from 'sinon';
import { CheckboxField } from '../CheckboxField';

describe('rendering', () => {
  it('renders correctly mandatory props only', () => {
    const tree = shallow(<CheckboxField id="test" label="label" />);

    expect(shallowToJson(tree)).toMatchSnapshot();
  });

  it('renders correctly with all props', () => {
    const tree = shallow(<CheckboxField
      checked
      disabled
      helpText="some help"
      id="test"
      inFormLayout
      isLabelVisible
      label="label"
      labelPosition="after"
      layout="horizontal"
      required
      value="value"
      validationState="warning"
      validationText="some error"
    />);

    expect(shallowToJson(tree)).toMatchSnapshot();
  });
});

describe('functionality', () => {
  it('calls onChange() when checked', () => {
    const spy = sinon.spy();
    const component = mount(<CheckboxField
      changeHandler={spy}
      id="test"
      label="label"
    />);

    component
      .find('input')
      .simulate('change', {
        preventDefault: () => {},
        target: { value: 'on' },
      });
    expect(spy.calledOnce).toEqual(true);
  });
});
