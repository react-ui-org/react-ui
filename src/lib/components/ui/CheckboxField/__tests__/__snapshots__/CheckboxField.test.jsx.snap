// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`rendering renders correctly mandatory props only 1`] = `
<div
  className="root   "
>
  <label
    className="inputWrap labelPositionAfter"
    htmlFor="test"
    id="test__label"
  >
    <input
      checked={false}
      className="input"
      disabled={false}
      id="test"
      onChange={null}
      required={false}
      type="checkbox"
    />
    <span
      className="label"
    >
      <span
        className="labelInner "
        id="test__labelText"
      >
        label
      </span>
    </span>
  </label>
</div>
`;

exports[`rendering renders correctly with all props 1`] = `
<div
  className="root isRootInFormLayout isRootLayoutHorizontal isRootStateWarning"
>
  <label
    className="inputWrap labelPositionAfter"
    htmlFor="test"
    id="test__label"
  >
    <input
      checked={true}
      className="input"
      disabled={true}
      id="test"
      onChange={null}
      required={true}
      type="checkbox"
      value="value"
    />
    <span
      className="label"
    >
      <span
        className="labelInner "
        id="test__labelText"
      >
        label
      </span>
    </span>
  </label>
  <div
    className="helpText"
    id="test__helpText"
  >
    some help
  </div>
  <div
    className="validationText"
    id="test__validationText"
  >
    some error
  </div>
</div>
`;
